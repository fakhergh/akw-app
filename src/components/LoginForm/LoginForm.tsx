import IconVisibility from '@mui/icons-material/Visibility';
import IconVisibilityOff from '@mui/icons-material/VisibilityOff';
import Button from '@mui/material/Button';
import Grid from '@mui/material/Grid';
import IconButton from '@mui/material/IconButton';
import InputAdornment from '@mui/material/InputAdornment';
import { Formik } from 'formik';
import { useCallback, useState } from 'react';

import { InputField } from '@/components/InputField/InputField';
import { BaseFormProps } from '@/interfaces/form';
import Yup from '@/utils/yup';

export interface LoginFormValues {
    email: string;
    password: string;
}

export interface LoginFormProps
    extends Omit<BaseFormProps<LoginFormValues>, 'onClose'> {
    userType: 'admin' | 'user';
}

const defaultValues: LoginFormValues = {
    email: '',
    password: '',
};

const adminDefaultValues: LoginFormValues = {
    email: 'admin@gmail.com',
    password: '0000',
};

const validationSchema = Yup.object().shape({
    email: Yup.string().email().required(),
    password: Yup.string().required(),
});

export function LoginForm({
    loading,
    onSubmit,
    userType,
    initialValues = userType === 'admin' ? adminDefaultValues : defaultValues,
}: LoginFormProps) {
    const [passwordVisible, setPasswordVisible] = useState(false);

    const togglePasswordVisibility = useCallback(() => {
        setPasswordVisible((prev) => !prev);
    }, []);

    return (
        <Formik
            initialValues={initialValues}
            onSubmit={onSubmit}
            validationSchema={validationSchema}
        >
            {({ handleSubmit }) => (
                <Grid container spacing={4}>
                    <Grid item xs={12}>
                        <InputField
                            data-testid="email-input"
                            margin="normal"
                            required
                            fullWidth
                            label="Email address"
                            name="email"
                            autoComplete="email"
                            disabled={loading}
                        />
                    </Grid>
                    <Grid item xs={12}>
                        <InputField
                            data-testid="password-input"
                            margin="normal"
                            required
                            fullWidth
                            name="password"
                            label="Password"
                            type={passwordVisible ? 'text' : 'password'}
                            autoComplete="current-password"
                            InputProps={{
                                endAdornment: (
                                    <InputAdornment position="end">
                                        <IconButton
                                            size="small"
                                            onClick={togglePasswordVisibility}
                                        >
                                            {passwordVisible ? (
                                                <IconVisibilityOff fontSize="small" />
                                            ) : (
                                                <IconVisibility fontSize="small" />
                                            )}
                                        </IconButton>
                                    </InputAdornment>
                                ),
                            }}
                            disabled={loading}
                        />
                    </Grid>
                    <Grid item xs={12}>
                        <Button
                            data-testid="submit-btn"
                            fullWidth
                            variant="contained"
                            sx={{ mt: 1, mb: 2 }}
                            onClick={() => handleSubmit()}
                            disabled={loading}
                            color={userType === 'admin' ? 'success' : 'primary'}
                        >
                            Login
                        </Button>
                    </Grid>
                </Grid>
            )}
        </Formik>
    );
}
