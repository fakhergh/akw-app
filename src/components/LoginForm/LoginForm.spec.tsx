import { render, waitFor } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import { LoginForm, LoginFormValues } from '@/components/LoginForm/LoginForm';

const values: LoginFormValues = {
    email: 'joe.hattab@gmail.com',
    password: '0000',
};

const mockOnSubmit = vi.fn();

describe('<LoginForm/>', () => {
    it('Should matches snapshot', () => {
        const { container } = render(
            <LoginForm userType="user" onSubmit={mockOnSubmit} />,
        );

        expect(container).toMatchSnapshot();
    });

    it('Should renders the form fields correctly', () => {
        const { getByTestId } = render(
            <LoginForm userType="user" onSubmit={mockOnSubmit} />,
        );

        expect(getByTestId('email-input')).toBeInTheDocument();
        expect(getByTestId('password-input')).toBeInTheDocument();
        expect(getByTestId('submit-btn')).toBeInTheDocument();
    });

    it('Should calls onSubmit with correct values when the form is submitted', async () => {
        const onSubmit = vi.fn();

        const { getByTestId } = render(
            <LoginForm userType="user" onSubmit={onSubmit} />,
        );

        const emailInput = getByTestId('email-input').querySelector('input');
        const passwordInput =
            getByTestId('password-input').querySelector('input');
        const submitButton = getByTestId('submit-btn');

        await userEvent.type(emailInput!, values.email);
        await userEvent.type(passwordInput!, values.password);
        await userEvent.click(submitButton);

        await waitFor(() => {
            expect(onSubmit).toHaveBeenCalledWith(values, expect.anything());
        });
    });
});
