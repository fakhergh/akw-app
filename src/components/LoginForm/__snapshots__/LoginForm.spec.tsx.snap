// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<LoginForm/> > Should matches snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-4z7gnm-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-ypts9r-MuiFormControl-root-MuiTextField-root"
        data-testid="email-input"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r0:"
          id=":r0:-label"
        >
          Email address
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-10kwqr4-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="email"
            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            name="email"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w1u3ce"
            >
              <span>
                Email address
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-ypts9r-MuiFormControl-root-MuiTextField-root"
        data-testid="password-input"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r1:"
          id=":r1:-label"
        >
          Password
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-1iq2yy6-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="current-password"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            name="password"
            required=""
            type="password"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-r363y4-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                data-testid="VisibilityIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5M12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5m0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w1u3ce"
            >
              <span>
                Password
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-13uxqmo-MuiButtonBase-root-MuiButton-root"
        data-testid="submit-btn"
        tabindex="0"
        type="button"
      >
        Login
      </button>
    </div>
  </div>
</div>
`;
